pennylane/capture/to_catalyst.py

```diff
@@ -0,0 +1,486 @@
+        converter.convert_plxpr_eqn(eqn)
+    converter.return_wires()
+    return converter.catalyst_xpr
+
+
+class CatalystConverter:
+    """A class that manages the conversion of pennylane variant qfunc jaxpr to catalyst variant jaxpr.
+
+    By using this class, we can manage the statefullness of the conversion and the clean up required
+    at the end.
+
+    This class is purely internal to the ``qfunc_jaxpr_to_catalyst`` helper function.
+
+    Stateful variables are:
+    * ``catalyst_xpr``
+    * ``_wire_map``: map from wire value to ``AbstractQbit`` produced by earlier operations on the wire
+    * ``_op_math_cache``: operators that are consumed by later equations
+    * ``count``: number of variables already existing.  Used when creating new variables.
+
+    Constants saved between calls are:
+    * ``_qreg``
+    * ``_num_device_wires``
+    * ``_shots``
+
+    """
+
+    def __init__(self, plxpr):
+
+        self._count = 0
+
+        self.catalyst_xpr = jax.core.Jaxpr(
+            constvars=plxpr.constvars,
+            invars=plxpr.invars,
+            outvars=[],
+            eqns=[],
+        )
+
+        self._qreg = None
+        self._wire_map = {}
+        self._op_math_cache = {}
+        self._num_device_wires = 0
+        self._shots = None
+
+    def _make_var(self, aval):
+        """Create a variable from an aval.
+
+        Side Effects:
+            Increments the ``_counts`` variable.
+        """
+        self._count += 1
+        return jax.core.Var(count=self._count, suffix="", aval=aval)
+
+    def _get_wire(self, orig_wire):
+        """Get the ``AbstractQubit`` corresponding to a given wire label.
+
+        If the wire has already been acted upon, this will retrieved the ``AbstractQbit`` stored
+        in ``self._wire_map``.  If the wire has not already been acted upon, it adds
+        a qubit extraction equation and returns the ``AbstractQbit`` extracted from the
+        register.
+
+        Side Effects:
+            Adds an extraction equation if needed.
+        """
+        wire = orig_wire.val
+        if wire in self._wire_map:
+            return self._wire_map[orig_wire.val]
+
+        wire_var = jax.core.Literal(
+            val=wire, aval=jax.core.ShapedArray(dtype=int, shape=(), weak_type=True)
+        )
+        invars = [self._qreg, wire_var]
+        wire = self._make_var(c_prims.AbstractQbit())
+        outvar = [wire]
+        qextract_eqn = jax.core.JaxprEqn(
+            invars,
+            outvar,
+            c_prims.qextract_p,
+            {},
+            effects=jax.core.no_effects,
+            source_info=null_source_info,
+        )
+        self.catalyst_xpr.eqns.append(qextract_eqn)
+        return wire
+
+    def add_device_eqn(self, device):
+        """Add the equation for setting a device.
+
+        Should be the first method called when populating the catalyst xpr.
+        """
+        self._num_device_wires = len(device.wires)
+        self._shots = device.shots
```

Comment:
Does this capture the scenario where a QNode received the shots kwarg? I suppose so, because the device temporarily has changed shots?